impress-org/givewp
# Feature: add persistence scoping to Fields

Resolves #6861 

## Description

This PR introduces key features for more flexible storage capabilities in the Field API. It's awesome.

The existing persistence options was inferring the name of meta from the field name, and choosing whether the meta was stored in the donation or donor meta.

First, this introduces a new concept called the "scope" to a Field. A scope can be any string, and is mosts likely to be `donation` or `donor`. When the form is being saved, it will know what to do based on those two scopes (i.e. store them in their respective meta). But if it's any other value, then an add-on can hook in at that point and choose what to do — whether it's store the value or simply have a side-effect.

As a note, the existing `shouldStoreAsDonorMeta` API is fully backwards-compatible, so nothing that's using those methods needs to change.

A `metaKey` can now also be set for the field, so the field name and meta key can be two separate things. We'll likely have the field name be the fallback for donation and donor storage.

## Affects

This is a new API with changes that are only under the hood, so it should be safe. It does affect how a field determines whether it's stored as part of the donor or donation.

## Testing Instructions

Just make sure that custom field meta is still stored in the right place. I believe this is used in the code that persists the old FFM fields.

## Pre-review Checklist






---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205225437697504
